planner/solver.go

```diff
@@ -333,21 +332,23 @@ func createPlanningSolutionCandidate(placeIndexes []int, placeClusters [][]match
 	return res, nil
 }
 
-func (s *Solver) FindBestPlanningSolutions(ctx context.Context, placeClusters [][]matching.Place, topSolutionsCount int, iterator *MultiDimIterator) (resp chan PlanningResp) {
+func (s *Solver) FindBestPlanningSolutions(ctx context.Context, placeClusters [][]matching.Place, topSolutionsCount int, iterator *MultiDimIterator) (resp *PlanningResp) {
 	if topSolutionsCount <= 0 {
 		topSolutionsCount = TopSolutionsCountDefault
 	}
 
 	priorityQueue := &MinPriorityQueue[Vertex]{}
 	// int8 is enough for place deduplication limit
 	includedPlaces := make(map[string]int8)
-	resp = make(chan PlanningResp, 1)
+	resp = &PlanningResp{}
 
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, SolverTimeout)
+	defer cancel()
 	for iterator.HasNext() {
 		select {
-		case <-ctx.Done():
-			resp <- PlanningResp{Err: errors.New(ComputationTimedOutErrMsg), ErrorCode: RequestTimeOut}
-			return
+		case <-ctxWithTimeout.Done():
```

Comment:
Since we changed the response under timeout, is it easy to add a unit test to cover this scenario, i.e. ctxWithTimeout?

Reply:
It is not easy to test, since the timeout happens inside the `FindBestPlanningSolutions` function. If you run locally you can see the error message when timeout happens.

Reply:
Sg, lgtm